src/kernel/pagecache.c

```diff
@@ -777,39 +839,103 @@ closure_function(1, 3, void, pagecache_read_sg,
     for (u64 pi = k.state_offset; pi < end; pi++) {
         if (pp == INVALID_ADDRESS || page_offset(pp) > pi) {
             pp = allocate_page_nodelocked(pn, pi);
             if (pp == INVALID_ADDRESS) {
-                pagecache_unlock_node(pn);
                 apply(apply_merge(m), timm("result", "failed to allocate pagecache_page"));
-                return;
+                break;
             }
-        } else if (page_state(pp) == PAGECACHE_PAGESTATE_FREE) {
-            pagecache_lock_state(pc);
-            realloc_pagelocked(pc, pp);
-            pagecache_unlock_state(pc);
         }
-
-        range r = byte_range_from_page(pc, pp);
-        range i = range_intersection(q, r);
-        u64 length = range_span(i);
-        sg_buf sgb = sg_list_tail_add(sg, length);
-        sgb->buf = pp->kvirt + (i.start - r.start);
-        sgb->size = length;
-        sgb->offset = 0;
-        sgb->refcount = &pp->refcount;
-        refcount_reserve(&pp->refcount);
-
-        touch_or_fill_page_nodelocked(pn, pp, m);
+        if (ph)
+            apply(ph, pp);
+        if (touch_page_locked(pn, pp, m)) {
+            /* This page does not need to be fetched: fetch pages accumulated so far in read_sg. */
+            if (read_sg) {
+                status_handler read_sh = apply_merge(m);
+                status_handler read_complete = closure(pc->h, pagecache_node_fetch_complete, pc,
+                    read_pp, range_span(read_r) >> pc->page_order, read_sg, read_sh);
+                if (read_complete == INVALID_ADDRESS) {
+                    apply(read_sh, timm("result", "failed to allocate read completion"));
+                    break;
+                }
+                pagecache_unlock_state(pc);
+                pagecache_debug("reading %R from node %p\n", read_r, pn);
+                apply(pn->fs_read, read_sg, read_r, read_complete);
+                pagecache_lock_state(pc);
+                read_sg = 0;
+            }
+        } else {
+            /* This page needs to be fetched: add it to read_sg. */
+            if (page_state(pp) == PAGECACHE_PAGESTATE_FREE) {
+                apply(apply_merge(m), timm("result", "failed to allocate page"));
+                break;
+            }
+            if (!read_sg) {
+                read_sg = allocate_sg_list();
+                if (read_sg == INVALID_ADDRESS) {
+                    apply(apply_merge(m), timm("result", "failed to allocate read SG list"));
+                    read_sg = 0;
+                    break;
+                }
+                read_pp = pp;
+                read_r = range_lshift(irangel(page_offset(pp), 0), pc->page_order);
+            }
+            pagecache_add_sgb(pc, pp, read_sg);
+            read_r.end += cache_pagesize(pc);
+        }
         pp = (pagecache_page)rbnode_get_next((rbnode)pp);
     }
+    pagecache_unlock_state(pc);
     pagecache_unlock_node(pn);
+    if (read_sg) {
+        status_handler read_sh = apply_merge(m);
```

Comment:
I think some refactoring could merge this fetch with the one above.

Reply:
Added a pagecache_node_fetch_sg() helper function to remove code duplication.